### ISSUE TITLE
after kops cluster upgrade cluster lost network connectivity

### ISSUE THREAD

**Issue by AndriiOmelianenko** [kubernetes/kops#7518]
**1. What `kops` version are you running? The command `kops version`, will display
 this information.**

Version 1.13.0 (git-be5fb9019)

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**

kubernetesVersion: 1.13.10
dockerVersion: 18.06.3
etcdVersion: 3.3.13

**3. What cloud provider are you using?**
aws

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**

```
apiVersion: kops/v1alpha2
kind: Cluster
metadata:
  name: redacted-redacted-prod.k8s.local
spec:
  additionalPolicies:
    master: |
      [
          redacted
        ]
    node: |
      [
          redacted
      ]
  api:
    loadBalancer:
      type: Internal
  authorization:
    alwaysAllow: {}
  channel: stable
  cloudProvider: aws
  configBase: s3://redacted-redacted-1-kubernetes-state/prod/redacted-redacted-prod.k8s.local/
  docker:
    liveRestore: true
    storage: overlay2
    storageOpts:
    - overlay2.override_kernel_check=true
    version: 18.06.3
  etcdClusters:
  - enableEtcdTLS: true
    encryptedVolume: true
    etcdMembers:
    - instanceGroup: redactedinfra_redacted_prod_k8s-master-a
      name: a
    - instanceGroup: redactedinfra_redacted_prod_k8s-master-b
      name: b
    - instanceGroup: redactedinfra_redacted_prod_k8s-master-c
      name: c
    name: main
    version: 3.3.13
  - enableEtcdTLS: true
    encryptedVolume: true
    etcdMembers:
    - instanceGroup: redactedinfra_redacted_prod_k8s-master-a
      name: a
    - instanceGroup: redactedinfra_redacted_prod_k8s-master-b
      name: b
    - instanceGroup: redactedinfra_redacted_prod_k8s-master-c
      name: c
    name: events
    version: 3.3.13
  fileAssets:
  - content: |
      apiVersion: audit.k8s.io/v1beta1
      kind: Policy
      rules:
        # The following requests were manually identified as high-volume and low-risk,
        # so drop them.
        - level: None
          users: ["system:kube-proxy"]
          verbs: ["watch"]
          resources:
            - group: "" # core
              resources: ["endpoints", "services", "services/status"]
        - level: None
          users: ["system:unsecured"]
          namespaces: ["kube-system"]
          verbs: ["get"]
          resources:
            - group: ""
              resources: ["configmaps"]
        - level: None
          users: ["kubelet"] # legacy kubelet identity
          verbs: ["get"]
          resources:
            - group: "" # core
              resources: ["nodes", "nodes/status"]
        - level: None
          userGroups: ["system:nodes"]
          verbs: ["get"]
          resources:
            - group: "" # core
              resources: ["nodes", "nodes/status"]
        - level: None
          users:
            - system:kube-controller-manager
            - system:kube-scheduler
            - system:serviceaccount:kube-system:endpoint-controller
          verbs: ["get", "update"]
          namespaces: ["kube-system"]
          resources:
            - group: "" # core
              resources: ["endpoints"]
        - level: None
          users: ["system:apiserver"]
          verbs: ["get"]
          resources:
            - group: "" # core
              resources: ["namespaces", "namespaces/status", "namespaces/finalize"]
        # Don't log HPA fetching metrics.
        - level: None
          users:
            - system:kube-controller-manager
          verbs: ["get", "list"]
          resources:
            - group: "metrics.k8s.io"
        # Don't log these read-only URLs.
        - level: None
          nonResourceURLs:
            - /healthz*
            - /version
            - /swagger*
        # Don't log events requests.
        - level: None
          resources:
            - group: "" # core
              resources: ["events"]
        # node and pod status calls from nodes are high-volume and can be large, don't log responses for expected updates from nodes
        - level: Request
          users: ["kubelet", "system:node-problem-detector", "system:serviceaccount:kube-system:node-problem-detector"]
          verbs: ["update","patch"]
          resources:
            - group: "" # core
              resources: ["nodes/status", "pods/status"]
          omitStages:
            - "RequestReceived"
        - level: Request
          userGroups: ["system:nodes"]
          verbs: ["update","patch"]
          resources:
            - group: "" # core
              resources: ["nodes/status", "pods/status"]
          omitStages:
            - "RequestReceived"
        # deletecollection calls can be large, don't log responses for expected namespace deletions
        - level: Request
          users: ["system:serviceaccount:kube-system:namespace-controller"]
          verbs: ["deletecollection"]
          omitStages:
            - "RequestReceived"
        # Secrets, ConfigMaps, and TokenReviews can contain sensitive & binary data,
        # so only log at the Metadata level.
        - level: Metadata
          resources:
            - group: "" # core
              resources: ["secrets", "configmaps"]
            - group: authentication.k8s.io
              resources: ["tokenreviews"]
          omitStages:
            - "RequestReceived"
        # Get repsonses can be large; skip them.
        - level: Request
          verbs: ["get", "list", "watch"]
          resources:
            - group: "" # core
            - group: "admissionregistration.k8s.io"
            - group: "apiextensions.k8s.io"
            - group: "apiregistration.k8s.io"
            - group: "apps"
            - group: "authentication.k8s.io"
            - group: "authorization.k8s.io"
            - group: "autoscaling"
            - group: "batch"
            - group: "certificates.k8s.io"
            - group: "extensions"
            - group: "metrics.k8s.io"
            - group: "networking.k8s.io"
            - group: "policy"
            - group: "rbac.authorization.k8s.io"
            - group: "settings.k8s.io"
            - group: "storage.k8s.io"
          omitStages:
            - "RequestReceived"
        # Default level for known APIs
        - level: RequestResponse
          resources:
            - group: "" # core
            - group: "admissionregistration.k8s.io"
            - group: "apiextensions.k8s.io"
            - group: "apiregistration.k8s.io"
            - group: "apps"
            - group: "authentication.k8s.io"
            - group: "authorization.k8s.io"
            - group: "autoscaling"
            - group: "batch"
            - group: "certificates.k8s.io"
            - group: "extensions"
            - group: "metrics.k8s.io"
            - group: "networking.k8s.io"
            - group: "policy"
            - group: "rbac.authorization.k8s.io"
            - group: "settings.k8s.io"
            - group: "storage.k8s.io"
          omitStages:
            - "RequestReceived"
        # Default level for all other requests.
        - level: Metadata
          omitStages:
            - "RequestReceived"
    name: audit-policy.conf
    path: /srv/kubernetes/audit-policy.conf
    roles:
    - Master
  - content: |
      fs.may_detach_mounts=1
    name: may_detach_mounts.conf
    path: /etc/sysctl.d/may_detach_mounts.conf
    roles: []
  hooks:
  - before:
    - docker.service
    manifest: |
      Type=oneshot
      ExecStart=/usr/sbin/sysctl -p /etc/sysctl.d/may_detach_mounts.conf
    name: systcl_setup.service
  kubeAPIServer:
    auditLogMaxAge: 1
    auditLogMaxBackups: 5
    auditLogMaxSize: 100
    auditLogPath: /var/log/kube-apiserver-audit.log
    auditPolicyFile: /srv/kubernetes/audit-policy.conf
    authorizationMode: RBAC
    enableAdmissionPlugins:
    - PersistentVolumeClaimResize
    featureGates:
      ExpandInUsePersistentVolumes: "true"
      ExpandPersistentVolumes: "true"
    oidcClientID: oidc-client.redacted-redacted-prod.k8s.local
    oidcGroupsClaim: groups
    oidcIssuerURL: redacted
    oidcUsernameClaim: email
    runtimeConfig:
      rbac.authorization.k8s.io/v1: "true"
  kubeControllerManager:
    featureGates:
      ExpandInUsePersistentVolumes: "true"
      ExpandPersistentVolumes: "true"
    terminatedPodGCThreshold: 10
  kubeDNS:
    provider: CoreDNS
  kubelet:
    anonymousAuth: false
    authenticationTokenWebhook: true
    authorizationMode: Webhook
    evictionHard: memory.available<5%,nodefs.available<10%,nodefs.inodesFree<10%,imagefs.available<10%,imagefs.inodesFree<10%
    evictionMinimumReclaim: memory.available=1Gi,nodefs.available=1Gi
    evictionSoft: memory.available<10%,nodefs.available<40%,nodefs.inodesFree<20%,imagefs.available<30%,imagefs.inodesFree<20%
    evictionSoftGracePeriod: memory.available=10m,nodefs.available=10m,nodefs.inodesFree=10m,imagefs.available=10m,imagefs.inodesFree=10m
    featureGates:
      ExpandInUsePersistentVolumes: "true"
      ExpandPersistentVolumes: "true"
    kubeReserved:
      cpu: 250m
      ephemeral-storage: 1Gi
      memory: 1Gi
    systemReserved:
      cpu: 150m
      ephemeral-storage: 1Gi
      memory: 1Gi
  kubernetesApiAccess:
  - 172.16.192.0/20
  - 172.16.224.0/22
  kubernetesVersion: 1.13.10
  masterPublicName: api.redacted-redacted-prod.k8s.local
  networkCIDR: 172.16.224.0/22
  networkID: vpc-redacted
  networking:
    weave:
      mtu: 8912
  nonMasqueradeCIDR: 100.64.0.0/10
  sshAccess:
  - 172.16.192.0/20
  - 172.16.224.0/22
  subnets:
  - cidr: 172.16.224.0/25
    id: subnet-redacted
    name: redacted-1a
    type: Private
    zone: redacted-1a
  - cidr: 172.16.225.128/25
    id: subnet-redacted
    name: redacted-1b
    type: Private
    zone: redacted-1b
  - cidr: 172.16.226.0/25
    id: subnet-redacted
    name: redacted-1c
    type: Private
    zone: redacted-1c
  - cidr: 172.16.227.0/26
    id: subnet-redacted
    name: utility-redacted-1a
    type: Utility
    zone: redacted-1a
  - cidr: 172.16.227.64/26
    id: subnet-redacted
    name: utility-redacted-1b
    type: Utility
    zone: redacted-1b
  - cidr: 172.16.227.128/26
    id: subnet-redacted
    name: utility-redacted-1c
    type: Utility
    zone: redacted-1c
  topology:
    dns:
      type: Private
    masters: private
    nodes: private
---
apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  labels:
    kops.k8s.io/cluster: redacted-redacted-prod.k8s.local
  name: redactedinfra_redacted_prod_k8s-master-a
spec:
  cloudLabels:
    Backup: "false"
    Environment: PRD
    OS: centos7
    Owner: kops
  image: ami-redacted
  machineType: m5.large
  maxSize: 1
  minSize: 1
  role: Master
  subnets:
  - redacted-1a
---
apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  labels:
    kops.k8s.io/cluster: redacted-redacted-prod.k8s.local
  name: redactedinfra_redacted_prod_k8s-master-b
spec:
  cloudLabels:
    Backup: "false"
    Environment: PRD
    OS: centos7
    Owner: kops
  image: ami-redacted
  machineType: m5.large
  maxSize: 1
  minSize: 1
  role: Master
  subnets:
  - redacted-1b
---
apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  labels:
    kops.k8s.io/cluster: redacted-redacted-prod.k8s.local
  name: redactedinfra_redacted_prod_k8s-master-c
spec:
  cloudLabels:
    Backup: "false"
    Environment: PRD
    OS: centos7
    Owner: kops
  image: ami-redacted
  machineType: m5.large
  maxSize: 1
  minSize: 1
  role: Master
  subnets:
  - redacted-1c
---
apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  labels:
    kops.k8s.io/cluster: redacted-redacted-prod.k8s.local
  name: redactedinfra_redacted_prod_k8s-node
spec:
  cloudLabels:
    Backup: "false"
    Environment: PRD
    OS: centos7
    Owner: kops
  image: ami-redacted
  machineType: m5.2xlarge
  maxSize: 6
  minSize: 6
  nodeLabels:
    k8snode-label: k8snode-label-value
  role: Node
  subnets:
  - redacted-1a
  - redacted-1b
  - redacted-1c
---
apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  labels:
    kops.k8s.io/cluster: redacted-redacted-prod.k8s.local
  name: redactedinfra_redacted_prod_k8s-node-redacted
spec:
  cloudLabels:
    Backup: "false"
    Environment: PRD
    OS: centos7
    Owner: kops
  image: ami-redacted
  machineType: m5.2xlarge
  maxSize: 10
  minSize: 10
  nodeLabels:
    k8snode-label: redacted
  role: Node
  subnets:
  - redacted-1a
  - redacted-1b
  - redacted-1c
  taints:
  - dedicated=redacted:NoSchedule
---
apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  labels:
    kops.k8s.io/cluster: redacted-redacted-prod.k8s.local
  name: redactedinfra_redacted_prod_k8s-node-redacted
spec:
  cloudLabels:
    Backup: "false"
    Environment: PRD
    OS: centos7
    Owner: kops
  image: ami-redacted
  machineType: r5.large
  maxSize: 2
  minSize: 2
  nodeLabels:
    k8snode-label: redacted
  role: Node
  subnets:
  - redacted-1a
  taints:
  - dedicated=redacted:NoSchedule
  zones:
  - redacted-1a

```
**4. What commands did you run?  What is the simplest way to reproduce this issue?**
**5. What happened after the commands executed?**
**6. What did you expect to happen?**

so we planned to do rolling upgrade from 1.12.X version to 1.13.10
(etcd also should be updated from 3.2.24 to 3.3.13)

we upgraded masters instance-groups one-by-one:
```
kops rolling-update cluster "${CLUSTER_NAME}" --instance-group=${ASG} -v 9 --master-interval=10m --yes
```

then we upgraded rest of the nodes (non-masters):
```
kops rolling-update cluster "${CLUSTER_NAME}" --master-interval=2m -v 9 --yes
```
```
I0829 22:05:50.240141     300 instancegroups.go:276] Cluster validated.
I0829 22:05:50.241035     300 rollingupdate.go:184] Rolling update completed for cluster "redacted-redacted-prod.k8s.local"!
```

Everything was good in the beginning; then we noticed some issues in the cluster logs.
After investigation, we discovered, that etcd was upgraded only in 1 master out of 3.
```
2019-08-30 10:55:14.976291 W | etcdserver: the local etcd version 3.2.24 is not up-to-date
2019-08-30 10:55:14.976309 W | etcdserver: member 3eac08706c25987e has a higher version 3.3.13
2019-08-30 10:55:18.979783 W | etcdserver: the local etcd version 3.2.24 is not up-to-date
2019-08-30 10:55:18.979803 W | etcdserver: member 3eac08706c25987e has a higher version 3.3.13
```
```
I0830 10:39:54.065353    2723 controller.go:417] mismatched version for peer peer{id:"etcd-b" endpoints:"172.16.225.144:3996" }: want "3.3.13", have "3.2.24"
I0830 10:39:54.065380    2723 controller.go:417] mismatched version for peer peer{id:"etcd-c" endpoints:"172.16.226.37:3996" }: want "3.3.13", have "3.2.24"
I0830 10:39:54.065411    2723 controller.go:517] detected that we need to upgrade/downgrade etcd
I0830 10:39:54.065417    2723 upgrade.go:126] doing in-place upgrade to "3.3.13"
I0830 10:39:54.065425    2723 upgrade.go:146] backing up cluster before in-place upgrade
W0830 10:39:54.071675    2723 controller.go:904] peer gave error while trying to do backup: rpc error: code = Unknown desc = LeadershipToken in request "5G-gVN5yeHPvkVWTeR4Q8g" is not current leader
W0830 10:39:54.072331    2723 controller.go:904] peer gave error while trying to do backup: rpc error: code = Unknown desc = LeadershipToken in request "5G-gVN5yeHPvkVWTeR4Q8g" is not current leader
I0830 10:39:54.079694    2723 backup.go:118] performing snapshot save to /tmp/885902996/snapshot.db.gz
I0830 10:39:57.124598    2723 vfssync.go:181] Copying data from /tmp/885902996/snapshot.db.gz to s3://redacted-redacted-1-kubernetes-state/prod/redacted-redacted-prod.k8s.local/backups/etcd/main/2019-08-30T10:39:57Z-004408/etcd.backup.gz
I0830 10:39:57.124628    2723 s3fs.go:128] Writing file "s3://redacted-redacted-1-kubernetes-state/prod/redacted-redacted-prod.k8s.local/backups/etcd/main/2019-08-30T10:39:57Z-004408/etcd.backup.gz"
I0830 10:39:57.647969    2723 s3fs.go:128] Writing file "s3://redacted-redacted-1-kubernetes-state/prod/redacted-redacted-prod.k8s.local/backups/etcd/main/2019-08-30T10:39:57Z-004408/_etcd_backup.meta"
I0830 10:39:57.702620    2723 backup.go:143] backup complete: name:"2019-08-30T10:39:57Z-004408"
I0830 10:39:57.707436    2723 controller.go:906] backup response: name:"2019-08-30T10:39:57Z-004408"
I0830 10:39:57.707459    2723 upgrade.go:151] backup done: name:"2019-08-30T10:39:57Z-004408"
I0830 10:39:57.707470    2723 upgrade.go:164] reconfiguring peer "etcd-c" from "3.2.24" -> "3.3.13"
W0830 10:39:57.708230    2723 controller.go:149] unexpected error running etcd cluster reconciliation loop: error reconfiguring etcd peer "etcd-c": rpc error: code = Unknown desc = LeadershipToken in request "5G-gVN5yeHPvkVWTeR4Q8g" is not current leader
```

so we decided to upgrade masters again, with `--cloud-only` option:
```
kops rolling-update cluster "${CLUSTER_NAME}" --instance-group=${ASG} -v 9 --force --cloudonly --master-interval=1s --yes
```

Masters were re-created, a little bit of k8s-api downtime, then it was up.
Then issues with networking started:

logs of weave-net (on all nodes):
```
WARN: 2019/08/30 13:34:28.496087 TCP connection from 172.16.227.112:2556 to 100.121.128.24:80 blocked by Weave NPC.
WARN: 2019/08/30 13:34:28.496100 TCP connection from 172.16.227.112:2544 to 100.124.128.23:80 blocked by Weave NPC.
WARN: 2019/08/30 13:34:28.496118 TCP connection from 172.16.227.112:2558 to 100.121.128.24:80 blocked by Weave NPC.
ERROR: logging before flag.Parse: E0830 13:34:29.056243 6848 reflector.go:205] github.com/weaveworks/weave/prog/weave-npc/main.go:322: Failed to list *v1.Pod: Get https://100.64.0.1:443/api/v1/pods?limit=500&resourceVersion=0: dial tcp 100.64.0.1:443: i/o timeout
ERROR: logging before flag.Parse: E0830 13:34:29.057255 6848 reflector.go:205] github.com/weaveworks/weave/prog/weave-npc/main.go:321: Failed to list *v1.Namespace: Get https://100.64.0.1:443/api/v1/namespaces?limit=500&resourceVersion=0: dial tcp 100.64.0.1:443: i/o timeout
ERROR: logging before flag.Parse: E0830 13:34:29.058270 6848 reflector.go:205] github.com/weaveworks/weave/prog/weave-npc/main.go:323: Failed to list *v1.NetworkPolicy: Get https://100.64.0.1:443/apis/networking.k8s.io/v1/networkpolicies?limit=500&resourceVersion=0: dial tcp 100.64.0.1:443: i/o timeout
WARN: 2019/08/30 13:34:29.497878 TCP connection from 172.16.227.112:2546 to 100.124.128.23:80 blocked by Weave NPC.
WARN: 2019/08/30 13:34:29.497904 TCP connection from 172.16.227.112:2560 to 100.121.128.24:80 blocked by Weave NPC.
WARN: 2019/08/30 13:34:29.497911 TCP connection from 172.16.227.112:2548 to 100.121.128.24:80 blocked by Weave NPC.
WARN: 2019/08/30 13:34:29.497919 TCP connection from 172.16.225.248:34942 to 100.121.0.24:50000 blocked by Weave NPC.
WARN: 2019/08/30 13:34:29.497925 TCP connection from 172.16.225.248:34938 to 100.121.0.24:50000 blocked by Weave NPC.
WARN: 2019/08/30 13:34:29.497931 TCP connection from 172.16.225.248:34934 to 100.121.0.24:50000 blocked by Weave NPC.
WARN: 2019/08/30 13:34:29.497937 TCP connection from 172.16.225.248:34930 to 100.121.0.24:50000 blocked by Weave NPC.
```
So basically weave-net (and also core-dns) could not connect to k8s-api, so all our services were unavailable for some time.
weave-net or core-dns pods recreations didn't help.

we fixed the issue by rebooting each non-master node one-by-one, only after that weave (and other services) connected to kube-api and continued serving traffic.

please let me know what or when we did wrong, if it was expected behavior, what to do to not stuck with this issue in the future, any help or guidance is appreciated!
